Support external Slurm DBD

hanwen-pcluste opened the pull request:
### Description of changes
This PR exposes a configuration parameter that allows the cluster to point to a slurmdbd residing at a location different than the head node of the cluster.

This PR provides an example CFN template that allow to create a slurmdbd stack external to the PC cluster.

### Tests
* Created a new unittest and new integration test. The new integration test has been passed

### References
* Link to impacted open issues.
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hehe7318 committed:
Setup ASG, NLB, TG, LT, cfn-hup hook to external slrumdbd

jdeamicis committed:
Add VPC and Subnet parameters

Add VPC and Subnet parameters to the stack (the VPC does
not work CDK as it is as it requires access to the account).

Add cfn-hup configuration file.

Change ports listened by slurmdbd stack to default slurmdbd
port 6819 (to be changed later).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

Add instance_type CFN parameter

Also, rename Subnet_id to SubnetId.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

Add management security groups for slurmdbd instance

Add pair of security groups to allow SSH access to the slurmdbd instance.

Add KeyName CFN parameter for the SSH connection.

Add key_name and security_group_ids to the launch template.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

Add IAM role and necessary policies. Reformat, define self.stack

Use ec2.CfnSecurityGroup instead of ec2.SecurityGroup

ec2.CfnSecurityGroup provides more flexibility

Signed-off-by: Hanwen <hanwenli@amazon.com>

Use CfnSecurityGroup instead of SecurityGroup for better flexibility

Signed-off-by: Hanwen <hanwenli@amazon.com>

hehe7318 committed:
Add CW log group. Restrict IAM policies resources. Add CFN param for cookbook. Add dna.json. Add entrypoints.

hehe7318 committed:
Add region in dna.json

hehe7318 committed:
Add stack_name in dna.json

hehe7318 committed:
Add dbma_database_name in cfn param and dna.json. Add nlb_dns_name and a boolean value if_external_slurmdbd in dna.json for cookbook.

hehe7318 committed:
Add user data for custom cookbook url

hehe7318 committed:
Address comments. Add pair of security groups for the slurm accounting traffic across between cluster head node and external slurmdbd instance via port 6819

jdeamicis committed:
Set up private IP for external slurmdbd instance

Add components to allow the assignment of static private secondary
IP address to the slurmdbd instance.

Add possibility to set up the external slurmdbd instance as a
standalone instance in the CDK app rather than using an ASG.

Minor improvements to the provisional user_data.sh file.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add Route53 DNS to external Slurmdbd stack

Add hosted zone belonging to the stack.

Add DNS record related to instance being generated by the
stack (only in the case of CfnInstance, not via the Auto
Scaling Group).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add check for private IP assignment

Add bash function to check that the secondary private IP was
correctly assigned to the instance via EC2.

Add DescribeNetworkInterfaces policy to allow the implemented
check.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add DBMS client SG to slurmdbd instance

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Make external slurmdbd user data modular

Add cloud-config portion of user_data.sh

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Fix cfn-init configuration for external slurmdbd

The configuration of the `commands` was wrong and therefore
the commands were not being run at instance creation.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add support for ExternalSlurmdbd in SlurmSettings

Add configuration parameter representing the location of an
external slurmdbd daemon.

Add validator to check that ExternalSlurmdbd and Database
(Slurm accounting) are not used together.

Add unit test for the validator.

Adapt test_stack_limit to new external slurmdbd parameter.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

Expose ExternalSlurmdbd section in cluster configuration file

Expose ExternalSlurmdbd configuration as a subschema with two sub-
parameters, namely `Host` and `Port`, which map to `AccountingStorageHost`
and `AccountingStoragePort` in the `slurm.conf`.

Adapt unit tests and full example config to new schema.

Fix missing `DatabaseName` field in full example config.

Fix minor issue in a comment in the code.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Allow traffic from external slurmdbd to slurmctld's

Add ingress rule in the external dbd client security group
to allow connections initiated from the slurmdbd.

Such connections may be established from the slurmdbd's side
when the slurmdbd drops and then recovers: in this case
it is the slurmdbd attempting to re-establish the connectivity
between itself and all the slurmctld's it was previously
connected to.

Remove useless egress rules from security groups (all outgoing
traffic is enabled by default at the moment).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Remove NLB and target group from external slurmdbd cdk app

The NLB + target group approach has been discarded due to cost
and connectivity issues.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Use static IP to populate slurmdbd's DbdAddr parameter

Export the IP into the instance dna.json.

Add minor comments to the code.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Use AWS-specific CFN Parameter types

This allows a minimal level of parameter validation at CFN level.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Disable Route53 setup in external slurmdbd cdk template

Leave this as commented code for future reference.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena committed:
Add outputs to ExternalSlurmDbd Stack

Add output values that should be used in cluster configuration
* SlurmDbdPrivateIP
* SlurmDbdPort
* AccountingClientSG
* SSHClientSG

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

jdeamicis committed:
Add S3 bucket to external slurmdbd stack

Add IAM policies to mount S3 bucket from the slurmdbd
instance.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

Use hash for name of S3 bucket in external slurmdbd stack

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena committed:
Add log_group_name to dna.json for ExternalSlurmDbd

This is required to properly configure the CW Agent.
This provides the name of the log_group where the agent
will create the log-streams and send the logs.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

jdeamicis committed:
Expose external slurmdbd S3 bucket as CFN stack output

Fix names of CFN outputs named after slurmdbd.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Make Slurmdbd port configurable in external slurmdbd stack template

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

hanwen-cluster committed:
[integ-tests] Simplify database_factory to database

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
[integ-tests] Create a fixture to create munge key

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
[integ-tests] Add integration tests for external Slurm DBD support

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Add Json template generated from Python AWS CDK code

This template will be provided in ParallelCluster public documentation and used in integration tests. This commit also adds a Github check to ensure the external-slurmdbd.json is updated everytime external_slurmdbd_stack.py is updated. This approach is consistent with what we are doing for "API Model Validation"

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
[integ-tests] Add --munge-key-secret-arn to use existing munge key

Signed-off-by: Hanwen <hanwenli@amazon.com>

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -0,0 +1,118 @@
+        return autoscaling.CfnAutoScalingGroup(
+            self,
+            "External-Slurmdbd-ASG",
+            max_size="1",
```
We should add a comment of why we want the ASG to have 1 as the min, max and desired capacity so that there is always one SlurmDBD instance

hanwen-pcluste replied:
The commit is outdated

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -2,25 +2,44 @@
+
+        # define networking stuff
+        self.vpc_id = CfnParameter(
+            self, "VPC_id", type="String", description="The VPC to be used for the Slurmdbd stack."
```
NitPick: Lets keep the Param name as CamelCase as done for all the others

hanwen-pcluste replied:
The commit is outdated

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -96,10 +172,16 @@ def _add_external_slurmdbd_load_balancer(
 
         # add listener to NLB
-        listener = nlb.add_listener("External-Slurmdbd-Listener", port=22)
+        listener = nlb.add_listener("External-Slurmdbd-Listener", port=6819)
```
Nitpick: Lets use a constant for the port as well, so that we dont have to change it at multiple places

himani2411 commented on cloudformation/external-slurmdbd/resources/user_data.sh:
```diff
@@ -0,0 +1,33 @@
+custom_cookbook_url="${CustomCookbookUrl}"
+
+if [ "${custom_cookbook_url}" != "NONE" ]; then
+  curl --retry 3 -v -L -o /etc/chef/aws-parallelcluster-cookbook.tgz ${custom_cookbook_url}
```
We should make the URL S3 compatible like we do for ParallelCluster Stack  too.

hanwen-pcluste replied:
Can we keep this simple because it is mainly used for development?

himani2411 replied:
Yes I agree that we use this for development. Though there have been instances where we have shared these settings to the customer.

Just in case if there are any future changes which requires us to provide a solution to the customer we can use these settings for an easier experience

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -61,42 +82,49 @@ def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+            "is_external_slurmdbd": True,
         }
 
         return {
```
Maybe we can add this as part of the Write_files section in user_data.sh

hanwen-pcluste replied:
Is there any benefit of using "Write_files". If not, we can keep the code as is

himani2411 replied:
Agreed. Infact I see that HN, CF and keep the dna_json data separate from user_data.sh

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -352,6 +386,12 @@ def _add_iam_role(self):
                     ),
+                    iam.PolicyStatement(
+                        actions=["ec2:AssignPrivateIpAddresses", "ec2:DescribeInstances"],
+                        resources=["*"],
```
I think this will be handled in the next commit. I am looking at the CR commit by commit

But the resource is way to permissive

hanwen-pcluste replied:
The resource cannot be limited because we don't know the instance id or network interfaces that the autoscaling group is going to launch

himani2411 commented on cloudformation/external-slurmdbd/resources/user_data.sh:
```diff
@@ -7,13 +7,21 @@ function vendor_cookbook
   export HOME="${!HOME_BAK}"
 }
 
+TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"`
```
Nitpick: There is a small time gap between bootsrapping the instance and when we wont have this 2nd IP address assigned to the instance. The user_data here is  not MIME part file so this is one of the last hook which will run before the instance is up and running for us


This is non blocking, however if we add the write_files section and make this a MIME file like we do for HN, CF,etc   
then adding/assigning the IP as part of the bootstrap should be possible.

hanwen-pcluste replied:
How much is the time gap?

himani2411 replied:
Its usually a few seconds, but not sure.

But do we need to check if we check if the IP is available( like a validation) or is that something we dont need to be responsible for?

hanwen-pcluste replied:
We don' check ip availability. "--allow-reassignment" just seize the ip address for the last call

hanwen-cluster committed:
Improve external Slurm dbd dna.json

By following the new structure, we can simplify cookbook code

Signed-off-by: Hanwen <hanwenli@amazon.com>

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -254,13 +252,13 @@ def _add_external_slurmdbd_launch_template(self):
             type="String",
```
NitPick: change the type to `AWS::EC2::KeyPair::KeyName`

himani2411 replied:
Ignore this

himani2411 commented on tests/integration-tests/tests/schedulers/conftest.py:
```diff
@@ -237,9 +237,14 @@ def test_resources_dir(datadir):
-    encoded_munge_key = base64.b64encode(random_key).decode("utf-8")
-    munge_key_secret_arn = store_secret_in_secret_manager(region, secret_string=encoded_munge_key)
-    yield encoded_munge_key, munge_key_secret_arn
+def munge_key(store_secret_in_secret_manager, region, request):
```
Non-Blocking: We should use this fixture for `test_custom_munge_key` too

himani2411 commented:
Where is the warning that we are supposed to provide about encryption?

hanwen-cluster committed:
Remove cfn-hup from Slurm dbd instance

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Use CloudFormation pseudo-parameter to get stack name

We want the stack name to be decoded at CloudFormation deploy time, instead of template generate time.

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Regenerate external-slurmdbd.json

This file is automatically generated from `cdk synth --no-version-reporting`

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Add a warning message to inform users about the unencrypted connections between cluster and external Slurm dbd

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-pcluste commented:
> Where is the warning that we are supposed to provide about encryption?

Thanks! I added another commit to add the warning message

himani2411 commented on tests/integration-tests/tests/schedulers/conftest.py:
```diff
@@ -119,46 +118,37 @@ def _create_database_stack(stack_factory, request, region, vpc_stack_for_databas
-
-    yield _database_factory
-
-    for region, stack_dict in created_database_stacks.items():
```
why remove the dictionary for this? Dont we need Region specific DB Stacks anymore?

hanwen-pcluste replied:
`@pytest.fixture(scope="class")` is region-specific by default

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -118,11 +118,14 @@ def _add_cfn_init_config(self):
+                "region": self.region,
+                "log_group_name": self._log_group.log_group_name,
+                "stack_name": self.stack_name,
+                "node_type": "ExternalSlurmDbd",
```
Nitpick: Do we need `is_external_slurmdbd` and NodeType both? If they are basically saying the same thing here

hanwen-pcluste replied:
Done. I removed `is_external_slurmdbd`

himani2411 commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -123,7 +123,7 @@ def _add_cfn_init_config(self):
                 "region": self.region,
                 "log_group_name": self._log_group.log_group_name,
-                "stack_name": self.stack_name,
+                "stack_name": Aws.STACK_NAME,
```
Ohh I learned something new!

himani2411 commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -197,3 +197,14 @@ def _validate(self, database, external_slurmdbd):
+    def _validate(self, external_slurmdbd):
+        if external_slurmdbd is not None:
+            self._add_failure(
+                "Traffic between ParallelCluster and the external Slurmdbd is not encrypted.",
```
Do we need to get this message approved?

hanwen-pcluste replied:
Done

hanwen-cluster committed:
[slurmdbd] Use region and account id in bucket name

The hash doesn't work because it is generated before the CloudFormation template is deployed.

Signed-off-by: Hanwen <hanwenli@amazon.com>
